homeassistant/components/valve/__init__.py

```diff
@@ -0,0 +1,261 @@
+        [ValveEntityFeature.OPEN | ValveEntityFeature.CLOSE],
+    )
+
+    return True
+
+
+async def async_setup_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Set up a config entry."""
+    component: EntityComponent[ValveEntity] = hass.data[DOMAIN]
+    return await component.async_setup_entry(entry)
+
+
+async def async_unload_entry(hass: HomeAssistant, entry: ConfigEntry) -> bool:
+    """Unload a config entry."""
+    component: EntityComponent[ValveEntity] = hass.data[DOMAIN]
+    return await component.async_unload_entry(entry)
+
+
+@dataclass
+class ValveEntityDescription(EntityDescription):
+    """A class that describes valve entities."""
+
+    device_class: ValveDeviceClass | None = None
+
+
+class ValveEntity(Entity):
+    """Base class for valve entities."""
+
+    entity_description: ValveEntityDescription
+    _attr_current_valve_position: int | None = None
+    _attr_device_class: ValveDeviceClass | None
+    _attr_is_closed: bool | None
+    _attr_is_closing: bool | None = None
+    _attr_is_opening: bool | None = None
+    _attr_supported_features: ValveEntityFeature = ValveEntityFeature(0)
+
+    _valve_is_last_toggle_direction_open = True
+
+    @property
+    def current_valve_position(self) -> int | None:
+        """Return current position of valve.
+
+        None is unknown, 0 is closed, 100 is fully open.
+        """
+        return self._attr_current_valve_position
+
+    @property
+    def device_class(self) -> ValveDeviceClass | None:
+        """Return the class of this entity."""
+        if hasattr(self, "_attr_device_class"):
+            return self._attr_device_class
+        if hasattr(self, "entity_description"):
+            return self.entity_description.device_class
+        return None
+
+    @property
+    @final
+    def state(self) -> str | None:
+        """Return the state of the valve."""
+        if self.is_opening:
+            self._valve_is_last_toggle_direction_open = True
+            return STATE_OPENING
+        if self.is_closing:
+            self._valve_is_last_toggle_direction_open = False
+            return STATE_CLOSING
+
+        if (closed := self.is_closed) is None:
+            return None
+
+        return STATE_CLOSED if closed else STATE_OPEN
+
+    @final
+    @property
+    def state_attributes(self) -> dict[str, Any]:
+        """Return the state attributes."""
+
+        return {ATTR_CURRENT_POSITION: self.current_valve_position}
+
+    @property
+    def supported_features(self) -> ValveEntityFeature:
+        """Flag supported features."""
+        return self._attr_supported_features
+
+    @property
+    def is_opening(self) -> bool | None:
+        """Return if the valve is opening or not."""
+        return self._attr_is_opening
+
+    @property
+    def is_closing(self) -> bool | None:
+        """Return if the valve is closing or not."""
+        return self._attr_is_closing
+
+    @property
+    def is_closed(self) -> bool | None:
```

Comment:
To simplify the logic, should we just remove this property and let a valve which only supports close/open report positions 0/100?
If we don't do that, please make sure this is enforced (from the [initial architecture discussion](https://github.com/home-assistant/architecture/discussions/971#discussioncomment-7136936)):

> One detail which we should not copy from cover is that it requires covers which know their position to implement an is_closed property; this should instead be calculated by the valve base class to enforce the definitions of open/closed in relation to the valve position.

Likewise, the open and close services should set the position if the valve supports setting position.

Reply:
It's weird for the platform to report a position if the device doesn't support position reporting. I think we should try and allow the platform to stay as close to the device features as possible.

We could add a separate final method that has the is closed calculation logic so that we don't risk platforms to overwrite it here. Not sure if we also need to raise if platforms implement both `is_closed` and `current_valve_position` or if we just ignore `is_closed` in that case.

Reply:
Maybe just put the logic in `state` as that's already final?

Reply:
> It's weird for the platform to report a position if the device doesn't support position reporting.

Yes, I agree it's a bit weird. As long it's not possible for a `valve` which supports an exact position to implement an `is_closed` property or `open` /  `close` methods I'm happy 👍

> Not sure if we also need to raise if platforms implement both is_closed and current_valve_position or if we just ignore is_closed in that case.

I think it might be annoying if we raise, integrations may want to share a class for for valves which do and do not support position. This will certainly be the case for template, mqtt, command line and similar integrations.

Reply:
@cibernox I want that we ensure two things to ensure the behavior and state is consistent so we avoid the issues in cover:
1. If a `valve` supports `current_valve_position`, the `is_closed` property should be ignored
2. If a `valve` sets `ValveEntityFeature.SET_POSITION`, calling services `valve.open` or `valve.close` should call `async_set_valve_position`

@MartinHjelmare when writing this, I realize *If a `valve` supports `current_valve_position`* is not very clear, there's no way to know if a valve supports it without an entity feature flag?

Reply:
Erik and me discussed `current_valve_position` vs `is_closed` and we suggest the following. Add a new property `reports_position` that must be set by valve entities. That way we can make sure that platforms do the right thing and only implements either `current_valve_position` or `is_closed`.

```py
_attr_reports_position: bool

@property
def reports_position(self) -> bool:
    """Return True if entity reports position, False otherwise."""
    if hasattr(self, "_attr_reports_position"):
        return self._attr_reports_position
    if hasattr(self, "entity_description"):
        return self.entity_description.reports_position
    raise ValueError(f"'reports_position' not set for {self.entity_id}.")

@property
def state(self) -> str | None:
    """Return the state of the valve."""
    reports_position = self.reports_position
    if self.is_opening:
        self._valve_is_last_toggle_direction_open = True
        return STATE_OPENING
    if self.is_closing:
        self._valve_is_last_toggle_direction_open = False
        return STATE_CLOSING
    if reports_position is True:
        if (current_valve_position := self.current_valve_position) is None:
            return None
        closed = return current_valve_position == 0
        return STATE_CLOSED if closed else STATE_OPEN
    if (closed := self.is_closed) is None:
        return None

    return STATE_CLOSED if closed else STATE_OPEN
```

Reply:
Point 2. from Erik above should also be addressed.

Reply:
I think I only have one question, and it's about testing. In order to have good coverage I have to test for intermediate states, like `opening` and `closing`.

However, in testing, it's not very clear to me how to simulate that. If I define a close a `set_valve_position` method like this:
```py
    def set_valve_position(self, position: int) -> None:
        if (position > self._attr_current_valve_position):
            self._attr_is_closing = False
            self._attr_is_opening = True
        else:
            self._attr_is_closing = True
            self._attr_is_opening = False
```
I can test that indeed the valve is now opening, but how do I "finish" the operation? I tried this:
```py
    _target_valve_position: int
    def set_valve_position(self, position: int) -> None:
        if (position > self._attr_current_valve_position):
            self._attr_is_closing = False
            self._attr_is_opening = True
        else:
            self._attr_is_closing = True
            self._attr_is_opening = False
        self._target_valve_position = position

    def finish_movement(self):
        self._attr_current_valve_position = self._target_valve_position
        self._attr_is_closing = False
        self._attr_is_opening = False
```

And then in testing when I want to test the final state I tried:
```py
assert is_closing(hass, ent2)
ent2.finish_movement()
assert is_closed(hass, ent2)
```
but it fails because it seems that the info in `hass.states` is not sync with the attribute changes I did on that method.

I may be approaching this completely wrong tho.

Reply:
The entity needs to update state when it wants the state machine to store a new state. Call `self.async_write_ha_state` in the entity.

Reply:
Thanks! That did the trick (along with some `await`)